hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/parse/UnsupportedOperationVisitor.java

```diff
@@ -311,12 +312,28 @@ private void processSelect(SqlSelect select) {
             throw unsupported(select.getGroup(), "GROUP BY");
         }
 
-        if (select.getFetch() != null) {
-            throw unsupported(select.getFetch(), "LIMIT");
+        // Check for nested fetch offset
+        select.accept(new SelectVisitor(select));
+    }
+
+    private static final class SelectVisitor extends SqlBasicVisitor<Void> {
```

Comment:
I do not think we need a visitor here. The `UnsupportedOperationVisitor` is already a visitor that traverses all nodes. It should be sufficient to log the first `SqlSelect` statement, which should be the top one, and then ensure that none of the other `SqlSelect` statements contain `FETCH/OFFSET`.

Reply:
Surprisingly the first `SqlSelect` being passed to the `validateQuery` is the nested one. This is why I had to implement a generic visitor. So, relying on the logic that the top-level select is being called the first is not reliable.